tests/rptest/tests/transactions_test.py

```diff
@@ -434,3 +444,132 @@ def get_tx_manager_broker():
+            timeout_sec=30,
+            backoff_sec=1)
+        # The receivers of transaction coordinator RPCs (the leader of the
+        # topic being written to and the group topic) should be on 'dst_node'.
+        wait_until(
+            lambda: admin.transfer_leadership_to(namespace="kafka",
+                                                 topic="__consumer_offsets",
+                                                 partition=0,
+                                                 target_id=dst_id),
+            timeout_sec=30,
+            backoff_sec=1)
+        wait_until(
+            lambda: admin.transfer_leadership_to(namespace="kafka",
+                                                 topic=self.output_t.name,
+                                                 partition=0,
+                                                 target_id=dst_id),
+            timeout_sec=30,
+            backoff_sec=1)
+
+        def run_txn(should_commit):
+            """
+            Runs through the steps of a transaction, using consumer groups. If
+            'should_commit' is True, commits the transactions, otherwise
+            aborts.
+
+            NOTE: this uses confluent_kafka, which doesn't do anything fancy in
+            the background, e.g. aggressively fetching coordinator leaders.
+            """
+            self.generate_data(self.output_t, self.max_records)
+            producer = ck.Producer({
+                'bootstrap.servers': self.redpanda.brokers(),
+                'transactional.id': str(self.txn_id),
+            })
+            producer.init_transactions()
+            consumer = ck.Consumer({
+                'bootstrap.servers': self.redpanda.brokers(),
+                'group.id': "test",
+                'auto.offset.reset': 'earliest',
+                'enable.auto.commit': False,
+            })
+
+            consumer.subscribe([self.output_t])
+            num_consumed_records = 0
+
+            producer.begin_transaction()
+            while num_consumed_records < self.max_records:
+                records = self.consume(consumer)
+                for r in records:
+                    producer.produce(self.output_t.name, r.value(), r.key())
+                num_consumed_records += len(records)
+
+            producer.send_offsets_to_transaction(
+                consumer.position(consumer.assignment()),
+                consumer.consumer_group_metadata())
+
+            if should_commit:
+                producer.commit_transaction()
+            else:
+                producer.abort_transaction()
+            producer.flush()
+            consumer.close()
+
+            self.txn_id += 1
+
+        run_txn(should_commit=True)
+        run_txn(should_commit=False)
+
+    @cluster(num_nodes=3)
+    def test_txn_rpcs_with_upgrade(self):
+        # Initialize a transaction to bootstrap our transaction internal topic.
+        producer = ck.Producer({
```

Comment:
If producer got timeout for transaction it will be fatal error before upgrade. So you should recreate it in this case